packages/SwingSet/src/kernel/vatTranslator.js

```diff
@@ -255,12 +258,20 @@ function makeTranslateVatSyscallToKernelSyscall(vatID, kernelKeeper) {
     let result = null;
     if (resultSlot) {
       insistVatType('promise', resultSlot);
-      result = mapVatSlotToKernelSlot(resultSlot);
+      // Require that the promise used by the vat to receive the result of this
+      // send is allocated by the vat if it does not support pipelining.
+      // When a promise previously received as the result of a delivery is used
+      // as the result of a send call, it's in effect doing a redirect, which
+      // is currently only supported for pipelining vats.
+      result = mapVatSlotToKernelSlot(resultSlot, {
+        requireVatAllocated: !enablePipelining,
```

Comment:
I think this also allows the case where the vat is re-using a vpid that it previously exported, and I'm not sure you want that (but I might be wrong, I need to think about it more carefully). The `allocatedByVat` flag is not a direct indicator of the history of the promise, although possible cases are probably limited (and limited even further by this new assertion).

We might be able to use `allocatedByVat` to achieve the goal, but maybe the new option should be more like "this must be a brand new promise"? (Note that I just filed #5189 and in that case we might see the vpid in `args` before we see it as `result`, which could cause a false negative for a "must be new" check.. we might salvage that by processing `result` before we translate `args`).

Reply:
Yeah, thinking about it more, I think the check should be `requireNew`.

When one of us fixes #5189, the new tests (where userspace includes the result promise in the args of the same message) will fail this `requireNew` check, and we'll fix it (both here and in liveslots) by translating the result before translating the arguments.

```js
const args = [];
const p = E(target).foo(args);
args.push(p);
```

Reply:
Happy to revert to the initial "fresh" / `requireNew` approach. After realizing that pipelining vats needed the flexibility, I no longer saw a reason to prevent regular vats from pre-allocating a promise (it is currently marked as supported in the docs).

> The `allocatedByVat` flag is not a direct indicator of the history of the promise

Just to make sure I understand this flag correctly, is there any way for `allocatedByVat` to be `true` if the promise was in fact imported and not exported? That's the looser restriction I had settled on after moving away from "fresh", which seemed to satisfy all documented usages (and tests).

Reply:
You're correct, if kernel is the one introducing the promise into the vat, then `allocatedByVat` will always be false. It's just that I avoid using "import" and "export" for Promises, since that's not nearly as meaningful as which side is the decider. For *objects*, import-vs-export is a big deal, but not so much for Promises.